### PR TITLE
feat(repl): add "--eval-file" flag to execute a script file on startup

### PR DESCRIPTION
**Pull request by najumancheril**
<h2>Syntax</h2>
deno repl --script /path/to/script

<h2>Purpose</h2>
The purpose of this change is to provide deno with an option that is analogous to python -i /path/to/source, which runs the specified source file as additional prelude and then shows the user a REPL that can be used to enter more code. Python's -i flag is very useful for prototyping and experimentation. It is also useful for students who want to set up example structures and then examine them in an interactive session to improve their understanding of the language. This feature is different from "deno run" because "deno run" exits once the specified module exits. deno run is not interactive.

<h2>Example</h2>

```
$ cat a.js
let a=5;

$ deno repl --script a.js
Deno 1.20.5
exit using ctrl+d or close()
> a
5
>

```

<h2>Related Issues</h2>
This feature (and closely related features) have been previously discussed:

1. #3173
2. #7425
3. #7938
4. #11317
5. #13256

<h2>Related Work</h2>
Related work includes the implementation of the --eval flag (https://github.com/denoland/deno/pull/11590) which allows the user to specify the prelude code as command-line parameters. This is useful, but may be inconvenient for larger preludes.

<h2>Potential Concerns</h2>

1. --script may be a bad flag name.  #13256 suggests the name --eval-file. --prelude is another possibility, but it may be confusing because cli/tools/repl/session.rs already specifies some initial commands in a PRELUDE variable.
2. There is discussion by @dsherret on #13256 that this feature may not be worth it. It is already possible to expose module exports efficiently via --eval. Do we want a way to expose local variables in an interactive session? I think the answer is "probably." For example, students may find it useful to use preludes which define variables in a global scope because they don't understand modules yet. Other users may find it useful during prototyping when the amount of code is small and not worth structuring as a module.
3. There is discussion by @dsherret on #11317 about possibly modifying the script code to rewrite module import paths, etc. This implementation does not attempt to do that.
4. If the user specifies both --script and --eval, which should be executed first? In this implementation, I execute --script first.
5. Perhaps we should support multiple prelude files? The current implementation supports only one.
6. Perhaps we should support http:// in addition to file system access?
7. This is my first Rust program, so there may be bugs or non-idiomatic patterns in the code.

### CODE REVIEW

File: cli/tools/repl/mod.rs [hunk 1]
```diff
@@ -58,9 +60,17 @@ async fn read_line_and_poll(
   }
 }
 
+fn read_script(script_path: String) -> Result<String, AnyError> {
+  let mut file = File::open(script_path)?;
+  let mut data = Vec::new();
+  file.read_to_end(&mut data)?;
+  Ok(String::from_utf8(data)?)
+}
```
Comment (bartlomieju) [inline cli/tools/repl/mod.rs:68]: We should reuse facilities from `FileFetcher` to provide ability to load remote scripts as well

**Review by bartlomieju — COMMENTED** [review]
This PR was discussed during last CLI design meeting and there's support for landing it. A few changes that were requested:

- the flag should be named `--eval-file`
- the flag should support multiple values
- both local and remote scripts should be supported
- loaded files should be type-stripped if they have `.ts` extension and left as they are for `.js`

@najumancheril this is a good start, we will need to add an integration test for this feature in `cli/tests/integration/repl_tests.rs`. Let me know if you need some pointers.

File: cli/tools/repl/mod.rs (hunk 1 above)
Comment (najumancheril) [reply cli/tools/repl/mod.rs:68]: This is great feedback, thank you. I will make these changes.

**Comment by najumancheril** [comment]
I have implemented the first three requested changes.
1) Flag is now named --eval-file.
2) Multiple value are now supported. I was not sure what syntax to use, so I copied the argument parsing code for allow-lists (e.g. deno repl --eval-file=./a.js,http://localhost:8000/b.js,http://localhost:8000/c.js). Note that equal sign is now required.
3) tools::repl::run takes a ProcState, which has a FileFetcher as member. I am using this to load scripts. It handles both local and remote scripts. It also respects --reload. Due to our use of deno_core::resolve_import, local script paths now need to start with ./, ../, or /.

I will look into type-stripping and tests tomorrow. Let me know if you have any feedback on the above.

**Review by bartlomieju — COMMENTED** [review]
@najumancheril this looks great! I think @dsherret should be able to provide some pointers regarding transpilation.

**Comment by najumancheril** [comment]
@bartlomieju and @dsherret, I don't think I need to make any additional changes to deno in order to transpile .ts script files passed via --eval-file into JavaScript. The REPL entry point that I am using to inject the script code (ReplSession.evaluate_line_and_get_output) already does TypeScript transpilation. Here is the sequence of events:

1. tools::repl::run loads each eval file's source code (JavaScript or TypeScript code) as a String and invokes ReplSession.evaluate_line_and_get_output on that code String
2. evaluate_line_and_get_output invokes ReplSession.evaluate_line_with_object_wrapping on the code
3. evaluate_line_with_object_wrapping invokes ReplSession.evaluate_ts_expression on the code
4. evaluate_ts_expression creates a ParsedSource object with the code and invokes .transpile(). **This does the TypeScript to JavaScript conversion.**
6. evaluate_ts_expression invokes ReplSession.evaluate_expression on the transpiled code,.

Does that explanation make sense? If so, then I will proceed with adding the integration tests.

**Comment by bartlomieju** [comment]
> @bartlomieju and @dsherret, I don't think I need to make any additional changes to deno in order to transpile .ts script files passed via --eval-file into JavaScript. The REPL entry point that I am using to inject the script code (ReplSession.evaluate_line_and_get_output) already does TypeScript transpilation. Here is the sequence of events:
> 
> 1. tools::repl::run loads each eval file's source code (JavaScript or TypeScript code) as a String and invokes ReplSession.evaluate_line_and_get_output on that code String
> 2. evaluate_line_and_get_output invokes ReplSession.evaluate_line_with_object_wrapping on the code
> 3. evaluate_line_with_object_wrapping invokes ReplSession.evaluate_ts_expression on the code
> 4. evaluate_ts_expression creates a ParsedSource object with the code and invokes .transpile(). **This does the TypeScript to JavaScript conversion.**
> 5. evaluate_ts_expression invokes ReplSession.evaluate_expression on the transpiled code,.
> 
> Does that explanation make sense? If so, then I will proceed with adding the integration tests.

Thanks, that makes sense. I wonder about edge cases, but let's leave that until tests are done (maybe tests will surface some of these cases). I'd like to land this PR as soon as possible, so we could ship it in Deno v1.21 on Wednesday.

**Comment by najumancheril** [comment]
@bartlomieju I have added several tests for the new --eval-file flag. Please let know if you'd like any additional tests and/or changes.

File: cli/flags.rs [hunk 2]
```diff
@@ -1299,6 +1309,16 @@ Ignore linting a file by adding an ignore comment at the top of the file:
...
   runtime_args(Command::new("repl"), false, true)
     .about("Read Eval Print Loop")
+    .arg(
+      Arg::new("eval-file")
+        .long("eval-file")
+        .min_values(1)
+        .takes_value(true)
+        .use_value_delimiter(true)
+        .require_equals(true)
+        .help("Evaluates the provided code file(s) when the REPL starts.")
```
Comment (bartlomieju) [inline cli/flags.rs:1319]:
```suggestion
        .help("Evaluates the provided file(s) as scripts when the REPL starts. Accepts file paths and URLs.")
```

**Review by bartlomieju — COMMENTED** [review]
Looks well implemented. I will try it locally and give it another review. @dsherret please take a look as well

**Comment by najumancheril** [comment]
I agree this is annoying. I just noticed that resolve_import returns a specific error code (ParseError::RelativeUrlWithoutBase) in this case, so I can check for that and retry with an adjusted eval file path. I will do this now.

**Comment by bartlomieju** [comment]
> I agree this is annoying. I just noticed that resolve_import returns a specific error code (ParseError::RelativeUrlWithoutBase) in this case, so I can check for that and retry with an adjusted eval file path. I will do this now.

I think you can just apply `resolve_url_or_path` function to each argument in that list.

**Comment by najumancheril** [comment]
OK I think we're good

**Comment by bartlomieju** [comment]
@najumancheril please merge `main` branch into your branch and fix lint errors to make CI happy.

**Review by dsherret — APPROVED** [review]
This seems reasonable to me and the implementation looks good! I think if someone provides a module to this, they'll just have to make sure they use absolute specifiers if they import anything.

**Comment by bartlomieju** [comment]
> This seems reasonable to me and the implementation looks good! I think if someone provides a module to this, they'll just have to make sure they use absolute specifiers if they import anything.

That's a good point. I think we should add more explanation for the `--eval-flag` help. Also we will need to update https://github.com/denoland/manual/blob/main/tools/repl.md @najumancheril could you look into that?

**Comment by najumancheril** [comment]
I have updated the deno manual with information about --eval-file: https://github.com/denoland/manual/pull/293

**Review by bartlomieju — APPROVED** [review]
LGTM

Thank you @najumancheril. Great work for a first contribution, I can see cool use cases for this feature.
